semcore/d3-chart/src/Pattern.tsx

```diff
@@ -0,0 +1,645 @@
+      size: [18, 18],
+      children: (
+        <>
+          <path fillRule='evenodd' d='M1.5 1.5h17v17h-17v-17Zm5 5v7h7v-7h-7Z' clipRule='evenodd' />
+        </>
+      ),
+    },
+  },
+  triangle: {
+    fill: {
+      viewBox: '0 0 21 20',
+      children: (
+        <>
+          <path d='m10 2 8.66 15H1.34L10 2Z' />
+        </>
+      ),
+    },
+    symbol: {
+      viewBox: '0 0 33 32',
+      size: [16, 16],
+      children: (
+        <>
+          <path d='m15.66 5 12.66 21.928H3L15.66 5Z' />
+        </>
+      ),
+    },
+  },
+  crescent: {
+    fill: {
+      viewBox: '0 0 21 20',
+      children: (
+        <>
+          <path d='M18.501 16.748c.269-.293.018-.748-.38-.748A9 9 0 0 1 11.466.94c.32-.35.13-.941-.345-.941-5.523 0-10 4.477-10 10s4.477 10 10 10a9.974 9.974 0 0 0 7.38-3.252Z' />
+        </>
+      ),
+    },
+    symbol: {
+      viewBox: '0 0 21 20',
+      size: [14.99, 17.14],
+      children: (
+        <>
+          <path d='M18.501 16.748c.269-.293.018-.748-.38-.748A9 9 0 0 1 11.466.94c.32-.35.13-.941-.345-.941-5.523 0-10 4.477-10 10s4.477 10 10 10a9.974 9.974 0 0 0 7.38-3.252Z' />
+        </>
+      ),
+    },
+  },
+  zigzag: {
+    fill: {
+      viewBox: '0 0 12 12',
+      children: (
+        <>
+          <path d='M6 4.621 1.379 0H5.62L6 .379 6.379 0h4.242L6 4.621ZM0 10.621 1.379 12H5.62L0 6.379v4.242ZM6.379 12h4.242L12 10.621V6.38L6.379 12Z' />
+        </>
+      ),
+    },
+    symbol: {
+      viewBox: '0 0 21 21',
+      size: [17.14, 16.02],
+      children: (
+        <>
+          <path d='m20.121 1.284-9.797 9.797L.121.88v8.485l10.203 10.203 9.797-9.797V1.284Z' />
+        </>
+      ),
+    },
+  },
+} satisfies Record<string, Pattern>;
+
+const defaultPatternsList = Object.values(defaultPatterns);
+const patternsSync = new WeakMap();
+patternsSync.set(defaultPatternsList, []);
+
+export type PatternsConfig = true | keyof typeof defaultPatterns | Pattern | Pattern[];
+
+const getPatternByKey = (patternKey: string, patternsConfig: PatternsConfig = true) => {
+  if (typeof patternsConfig === 'object' && patternsConfig && !Array.isArray(patternsConfig))
+    return patternsConfig;
+  if (typeof patternsConfig === 'string') {
+    if (!defaultPatterns[patternsConfig]) {
+      const available = Object.keys(defaultPatterns).join(', ');
+      throw new Error(
+        `No default pattern ${patternsConfig} found. Options ${available} are available. If you want to use cusom pattern, pass it as object.`,
+      );
+    }
+    return defaultPatterns[patternsConfig];
+  }
+  const patterns = patternsConfig === true ? defaultPatternsList : patternsConfig;
+
+  if (!patternsSync.has(patterns)) {
+    patternsSync.set(patterns, []);
+  }
+  const patternsSyncList = patternsSync.get(patterns);
+  let index = patternsSyncList.indexOf(patternKey);
+  if (index === -1) {
+    index = patternsSyncList.length;
+    patternsSyncList.push(patternKey);
+  }
+  const patternIndex = index % patterns.length;
```

Comment:
looks like to difficult method... and it is not `getSomething`. It is get and sometimes store and magic =)
I suggest:
1) let's rewrite this place
```
const patternsSyncList = patternsSync.get(patterns);
const index = patternsSyncList.indexOf(patternKey);
if (index === -1) {
    patternsSyncList.push(patternKey);
    return patterns[0];
}
return patterns[index];
```
This is optional - just for think about it.
2) Maybe it is better to make a patternsList (default) some object (Map) with method to add some new patterns in it? And it will be obvious manually action - addSomeNewPattern. And getter will be just getter.

Reply:
> let's rewrite this place

Ok, that makes sense that this attempts was too messy. 
1. I renamed the function to `resolvePattern`.
2. I wrapped all the things around the `WeakMap` into function called `resolvePatternIndex`.
3. The `WeakMap` instead of _syncingThing_  is not called `patternsKeysOrderCache` to make it more clear why it exists.

> Maybe it is better to make a patternsList (default) some object (Map) with method to add some new patterns in it? And it will be obvious manually action - addSomeNewPattern. And getter will be just getter.

That was how the first approach was made. But actually this has not so much sense. 

Default patterns list allows developers enable patterns by just passing `patterns={true}` our of the box. If develops want to set custom pattern, most likely that they want to set it on specific parts of the chart instead of adding it to the list that is used for automatic patterns distributing and may affect other charts behavior.